src/shogun/machine/EnsembleMachine.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Yuhui Liu
+ */
+#ifndef __ENSEMBLEMACHINE_H
+#define __ENSEMBLEMACHINE_H
+
+#include <algorithm>
+#include <memory>
+#include <shogun/ensemble/CombinationRule.h>
+#include <shogun/features/Features.h>
+#include <shogun/labels/Labels.h>
+#include <shogun/machine/Machine.h>
+#include <shogun/util/traits.h>
+#include <shogun/util/zip_iterator.h>
+#include <vector>
+namespace shogun
+{
+
+	class EnsembleMachine : public SGObject
+	{
+	public:
+		EnsembleMachine() = default;
+        
+		EnsembleMachine(const EnsembleMachine& other)
+		    : m_machines(other.m_machines)
+		{
+		}
+
+		template <
+		    typename T,
+		    typename = std::enable_if_t<traits::is_container<T>::value>>
+		EnsembleMachine(const T& container)
+		{
+			std::copy(
+			    container.begin(), container.end(),
+			    std::back_inserter(m_machines));
+		}
+		~EnsembleMachine() = default;
+
+		void
+		set_combination_rule(std::shared_ptr<CombinationRule> combination_rule)
+		{
+			m_combination_rule = combination_rule;
+		}
+
+		void add_machine(std::shared_ptr<Machine> machine)
+		{
+			m_machines.push_back(machine);
+		}
+
+		void train(
+		    const std::shared_ptr<Features>& data,
+		    const std::shared_ptr<Labels>& labs)
+		{
+			for (const auto& machine : m_machines)
```

Comment:
does this compile? pretty sure it cannot be `const`?

Reply:
yes, the compiler doesn't complain anything, it can not be `const` because use `set_label` in here?

Reply:
ah nvm, it's the `shared_ptr` that is `const`, not the underlying type